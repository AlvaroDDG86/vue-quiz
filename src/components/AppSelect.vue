<template>
  <div class="app-select" @blur="blurHandler">
    <div
      class="app-select__selected"
      :class="{ open: open }"
      @click="clickSelectHandler"
    >
      <span v-if="selected.description">
        {{ selected.description }}
      </span>
      <span v-else-if="placeholder">{{ placeholder }}</span>
    </div>
    <div
      class="app-select__items"
      :class="{ 'app-select__items--close': !open }"
    >
      <div
        class="app-select__item"
        v-for="(option, i) of options"
        :key="i"
        @click="clickItemHandler(option)"
      >
        {{ option.description }}
      </div>
    </div>
  </div>
</template>
<script lang="ts">
import { defineComponent, PropType, ref, toRefs } from "vue";
import { SelectItem } from "@/models/select.model";
export default defineComponent({
  name: "AppSelect",
  props: {
    placeholder: {
      type: String,
      default: "",
    },
    options: {
      type: Object as PropType<SelectItem[]>,
      require: true,
    },
  },
  setup(props, { emit }) {
    const { options: optionsList } = toRefs(props);
    const selected = ref<SelectItem>({} as SelectItem);
    const open = ref<boolean>(false);

    const blurHandler = () => {
      open.value = false;
    };
    const clickSelectHandler = () => {
      open.value = !open.value;
    };
    const clickItemHandler = (option: SelectItem) => {
      selected.value = option;
      open.value = false;
      emit("input", option);
    };

    return {
      open,
      selected,
      optionsList,
      blurHandler,
      clickSelectHandler,
      clickItemHandler,
    };
  },
});
</script>

<style lang="postcss" scoped>
.app-select {
  @apply relative w-full text-center outline-none py-2 px-4;
  &__selected {
    @apply bg-gray-100 text-gray-800 cursor-pointer select-none rounded-lg flex justify-between items-center py-4 px-1 h-10;
    @apply dark:bg-gray-600 dark:text-gray-100;
  }

  &__items {
    @apply bg-gray-100 text-gray-800 overflow-hidden absolute left-0 right-0 z-10 shadow-xl mx-4 my-2 rounded-lg;
    @apply dark:bg-gray-600 dark:text-gray-100;

    &--close {
      @apply hidden;
    }
  }

  &__item {
    @apply cursor-pointer py-2;
    @apply hover:bg-gray-400;
  }
}
</style>
