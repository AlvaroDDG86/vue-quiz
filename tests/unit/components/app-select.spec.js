import { shallowMount } from "@vue/test-utils";
import AppSelect from "@/components/AppSelect";

describe("Question", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(AppSelect, {
      propsData: {
        options: [{ id: 1, description: "Option 1" }],
        placeholder: "placeholder test",
      },
    });
  });
  test("Should match snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  test("should display the placeholder", () => {
    expect(wrapper.find(".app-select__selected").text()).toBe(
      "placeholder test"
    );
  });
  test("should display the options", () => {
    wrapper.find(".app-select__selected").trigger("click");
    expect(wrapper.find(".app-select__item").exists()).toBe(true);
  });
  test("should select an option and emit selected value", () => {
    wrapper.find(".app-select__selected").trigger("click");
    wrapper.find(".app-select__item").trigger("click");
    expect(wrapper.emitted("input").length).toBe(1);
    expect(wrapper.emitted("input")[0][0]).toEqual({
      id: 1,
      description: "Option 1",
    });
  });
});
