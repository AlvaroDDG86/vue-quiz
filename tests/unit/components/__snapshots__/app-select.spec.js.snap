// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppSelect Should match snapshot 1`] = `
<div class="app-select">
  <div class="app-select__selected"><span>placeholder test</span></div>
  <div class="app-select__items app-select__items--close">
    <div class="app-select__item">Option 1</div>
  </div>
</div>
`;
